### ISSUE TITLE
Provide framework for decorating tests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2906]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2906|
|Type|enhancement|
|Created|2007-11-16 19:43:08Z|
|Branch|https://github.com/twisted/twisted/tree/itestcase-decorator-2906-2|

We should provide a way to decorate all tests in a test suite, preserving the structure of the test suite.

The first obvious case for this is using `ITestCase` as the decorator.

<details><summary>Searchable metadata</summary>

```
trac-id__2906 2906
type__enhancement enhancement
reporter__jml jml
priority__highest highest
milestone__ 
branch__branches_itestcase_decorator_2906_2 branches/itestcase-decorator-2906-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1195242188000000 1195242188000000
changetime__1198363214000000 1198363214000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

(In [21828]) Branching to 'itestcase-decorator-2906'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

(In [21829]) Apply changes from a local branch. Refs [#2906](https://github.com/twisted/twisted/issues/2906)

 * Add a 'decorate' method to unittest.
 * Add a result adapter that decorates all of its tests.
 * Change the test runner to decorate all tests with ITestCase, moving the
   point of compatibility to runtime and probably fixing issues affecting
   test suite hooks.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml commented|
|-|-|

A mere 694 lines of diff:

```
 reporter.py           |   83 ++++++++++++++++++++++++++++++
 runner.py             |   65 ++----------------------
 test/test_loader.py   |    4 +
 test/test_reporter.py |  134 ++++++++++++++++++++++++++++++++++++++++++++++++--
 test/test_runner.py   |    9 ++-
 test/test_tests.py    |   97 ++++++++++++++++++++++++++++++++++++
 unittest.py           |   99 ++++++++++++++++++++++++++----------
 7 files changed, 399 insertions(+), 92 deletions(-)
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

Pyflakes errors:
```
pyflakes errors:
twisted/trial/test/test_loader.py:170: redefinition of unused 'reporter' from line 8
twisted/trial/runner.py:25: 'suiteVisit' imported but unused
twisted/trial/runner.py:666: 'readline' imported but unused
```

I have two errors under Python 2.3:
```
===============================================================================
[FAIL]: twisted.trial.test.test_reporter.TestErrorReporting.testDoctestError

Traceback (most recent call last):
  File "/home/th/twisted_dev/twisted/trunk/twisted/trial/test/test_reporter.py", line 151, in testDoctestError
    self.stringComparison(expect, output.splitlines())
  File "/home/th/twisted_dev/twisted/trunk/twisted/trial/test/test_reporter.py", line 55, in stringComparison
    "Line %d: %r did not match string %r"
twisted.trial.unittest.FailTest: Line 1: '\\[(ERROR|FAIL)\\]: .*twisted\\.trial\\.test\\.erroneous\\.unexpectedException' did not match string '[ERROR]: run
it'
===============================================================================
[FAIL]: twisted.trial.test.test_reporter.TestUncleanWarningWrapperErrorReporting.testDoctestError

Traceback (most recent call last):
  File "/home/th/twisted_dev/twisted/trunk/twisted/trial/test/test_reporter.py", line 151, in testDoctestError
    self.stringComparison(expect, output.splitlines())
  File "/home/th/twisted_dev/twisted/trunk/twisted/trial/test/test_reporter.py", line 55, in stringComparison
    "Line %d: %r did not match string %r"
twisted.trial.unittest.FailTest: Line 1: '\\[(ERROR|FAIL)\\]: .*twisted\\.trial\\.test\\.erroneous\\.unexpectedException' did not match string '[ERROR]: run
it'
```


I guess that's all I have for now, because I don't really understand these changes :/. I'll try harder if someone doesn't beat me to it.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml commented|
|-|-|

I've corrected these failures, I think, and have fixed up the pyflakes that represented actual errors.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @jml|
|-|-|

The branch needs to be merged forward.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_9"></a>|@jml commented|
|-|-|

(In [22093]) Branching to 'itestcase-decorator-2906-2'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_10"></a>|@jml removed owner|
|-|-|

Merged forward.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [22097]) Merge r22094 from trunk, which was intended to go there

refs [#2906](https://github.com/twisted/twisted/issues/2906)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @jml|
|-|-|

 * proxyForInterface has moved, and there is a call from the old place in unittest.py
 * the way doctests are presented in the reporter is broken. I've seen it running the web2 test suite (twisted.web2.test.test_stream). Instead of having doctest -> ```DocTestCase```, I have twisted.trial.unittest -> _BrokenIDTestCaseAdapter. The current behavior isn't so great but a bit more clear.
 * ```unittest.TestDecorator``` methods are undocumented.

Overall, that looks OK, but I still don't understand what you want to do. You're talking about the obvious case of ```ITestCase```, can you elaborate a bit? Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_13"></a>|@jml commented|
|-|-|

There are many times when we want to globally change the behaviour of tests in some well defined way. Given that tests are largely indivisible lumps of code, a sane way to do this is to _decorate_ each test with a wrapper that provides this change of behaviour.

Two examples:

 1. `ITestCase`. Given that we want to support Python standard library unittests and given that Trial requires a slightly different set of behaviours from its tests, it makes sense to decorate all of the loaded unit tests with `ITestCase`. This is a no-op for Twisted tests and calls an adapter for standard library tests.
 1. _Forcing garbage collection_. Sometimes we want to force garbage collection before and after every test. Currently, we do that using special logic _inside_ `TestCase.run`. This is silly, because garbage collection is something independent of how a test runs. Far better to do it by decorating every single unittest with something that runs `gc.collect(); TestCase.run(); gc.collect()`.

Does that make it clearer?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

Alright! Indeed I understand the goal of the ticket much better.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_15"></a>|@jml set owner to @therve|
|-|-|

r22119 r22120

Done.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set owner to @jml|
|-|-|

The last docstrings use :param instead of @param. Everything else looks good, so please merge when corrected.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_17"></a>|@jml set status to `closed`|
|-|-|

(In [22125]) Use a decorator and adapter system to emnsure test compatibility.

 * Author: jml
 * Reviewer: therve
 * Fixes [#2906](https://github.com/twisted/twisted/issues/2906)

Adds an adapter to ITestCase from the standard library TestCase and adapts
all tests to ITestCase.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
